Bluetooth: Auto-update LE data length to max. supported

cvinayak opened the pull request.

cvinayak committed:
Bluetooth: controller: Fix DLE crossover assert

When initiating Data Length Change at the same instant the
crossover condition was not handled correctly causing the
controller to assert.

This fix will allow crossover of Data Length Update
procedure, and this collison is harmless as per Bluetooth
specification, and gracefully handled by the controller.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Auto-update LE data length to max. supported

Added implementation to auto-update LE Data Length to max.
Tx octets supported by the local and peer controllers.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mike-scott commented:
Tested-by: Michael Scott <michael.scott@linaro.org>

Tested successfully on nRF52382 for regressions in isr_rx_conn_pkt_ctrl_dle() function checks of eff_tx_octets and eff_rx_octets.  Didn't test new speed change functionality specifically.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -802,6 +834,13 @@ static void le_conn_complete(struct net_buf *buf)
 		}
 	}
 
+	if (BT_FEAT_LE_DLE(bt_dev.le.features)) {
```
How come you're doing this both here and in le_remote_feat_complete? Wouldn't you want to wait until you know if the remote supports the feature or not?

carlescufi replied:
Agreed, this check should be in the Host in my opinion.

cvinayak replied:
As a peripheral, if remote does not support slave initiated feature request, them this tries the DLE procedure; if peer does not support it will respond with unknown rsp.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -834,6 +873,15 @@ static void le_remote_feat_complete(struct net_buf *buf)
 		       sizeof(conn->le.features));
 	}
 
+	if (BT_FEAT_LE_DLE(bt_dev.le.features)) {
```
Don't you also want to check if the remote supports the feature, in addition to the local controller supporting it?

carlescufi replied:
Agreed, this check should be in the Host in my opinion.

cvinayak replied:
As a peripheral if peer does not support slave initiated feature request, we still want to check DLE is supported.

cvinayak replied:
agree here.

carlescufi commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -849,6 +897,41 @@ static void le_remote_feat_complete(struct net_buf *buf)
+	    BT_FEAT_LE_PHY_CODED(bt_dev.le.features)) {
+		int err;
+
+		err = hci_le_set_phy(conn);
```
Aren't you setting the PHY upon connection establishment too?

cvinayak replied:
No, legacy connection is established in 1M, then the switch has to happen.In LE Adv Ext, I think it can be established on the desired PHY on initiation.

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -802,6 +834,13 @@ static void le_conn_complete(struct net_buf *buf)
+		}
+	}
+
 	if (BT_FEAT_LE_PHY_2M(bt_dev.le.features) ||
```
This I will fix in a different patch, to check remote support.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -802,6 +834,13 @@ static void le_conn_complete(struct net_buf *buf)
+	if (BT_FEAT_LE_DLE(bt_dev.le.features)) {
+		err = hci_le_set_data_len(conn);
+		if (!err) {
+			goto done;
```
Just checking: is this "if (!err) goto done;" logic correct for the DLE and set_phy() branches? The intention of it for the preceding read_remote_features() branch is that we continue the remaining steps once we get the remote features complete event.

cvinayak replied:
yes, if err  == 0, then wait for data_len_change event to arrive and proceed to setting phy in that event function.

cvinayak commented:
@jhedberg @carlescufi do not approve/merge; data length change event being not a response event (like remote feature response), could arrive many times and the le_data_len_change function could get run many time. I will need to fix this PR so that le_data_len_change function that initiates PHY update gets executed only once.

carlescufi commented:
This requires further tests with an iPhone 7 and an air sniffer, since there are clearly issues. I would postpone this until we have been able to run such tests.

cvinayak committed:
Bluetooth: controller: Handle Rej Ext Ind for Length Req PDU

Added implementation to handle an incoming Reject Ext Ind PDU in
response to a sent Length Req PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
